tools/run_tests/xds_k8s_test_driver/tests/authz_test.py

```diff
@@ -0,0 +1,313 @@
+                    "ports": [1, self.server_port, 65535],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "host-match2",
+                    },
+                }],
+            },
+            {
+                "destinations": {
+                    "hosts": [
+                        f"not-the-host:{self.server_xds_port}",
+                        "not-the-host",
+                    ],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-host",
+                    },
+                },
+            },
+            {
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [1],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-port",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": ["*"],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "principal-present",
+                    },
+                },
+            },
+            {
+                "sources": [{
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                }, {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                        f"spiffe://{self.project}.svc.id.goog/ns/"
+                        f"{self.client_namespace}/sa/{self.client_name}",
+                    ],
+                }],
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "match-principal",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-principal",
+                    },
+                },
+            },
+        ]
```

Comment:
optional: Based on the test case content, we might be able to split the Authz policy into multiple chucks, so each test case only configures its minimum required policy then tests it. This might help people to write more tests or debug in future.

Reply:
Speaking of unexpected green runs. Is Authz config visible in the xDS configs? Do you think we should add a section to validate the received xDS config from control plane?

Reply:
> Is Authz config visible in the xDS configs?

Yep. It is an http filter, and is showing up in the Listener. I have a dump I can show you if you are interested.

> Do you think we should add a section to validate the received xDS config from control plane?

I don't know how much of that is being done elsewhere, but it just asserting the config matches a golden? 

The xds representation is pretty verbose because of and/or. Here is a single rule represented in xDS.

```
                          "ejona-psm-test-authz-policy-20211007-1801-n9leh_td_rbac_rule_7": {
                            "permissions": [
                              {
                                "andRules": {
                                  "rules": [
                                    {
                                      "orRules": {
                                        "rules": [
                                          {
                                            "header": {
                                              "name": ":authority",
                                              "suffixMatch": ":27468"
                                            }
                                          }
                                        ]
                                      }
                                    },
                                    {
                                      "orRules": {
                                        "rules": [
                                          {
                                            "destinationPort": 8080
                                          }
                                        ]
                                      }
                                    },
                                    {
                                      "orRules": {
                                        "rules": [
                                          {
                                            "header": {
                                              "name": "test",
                                              "safeRegexMatch": {
                                                "googleRe2": {
null
                                                },
                                                "regex": "never-match-principal"
                                              }
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ],
                            "principals": [
                              {
                                "andIds": {
                                  "ids": [
                                    {
                                      "orIds": {
                                        "ids": [
                                          {
                                            "authenticated": {
                                              "principalName": {
                                                "exact": "spiffe://foo.svc.id.goog/not/the/client"
                                              }
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
```

Reply:
> it just asserting the config matches a golden?

Yep, we can match the content of an HTTPFilter config with a predefined constant. It's done in every url-map tests, and most new GKE tests.